File: surface/widgets/nativescript/components/BindingProvider.ts
// ============================================================
// Clef Surface NativeScript Widget — BindingProvider
//
// Manages Clef Surface concept binding in NativeScript context.
// Provides connection state management and signal-based data
// flow via a global context pattern.
// ============================================================

import {
  StackLayout,
  Label,
  Color,
  Observable,
} from '@nativescript/core';

import type {
  BindingConfig,
  BindingMode,
  Signal,
} from '../../shared/types.js';

// --------------- Types ---------------

export type ConnectionState = 'disconnected' | 'connecting' | 'connected' | 'error' | 'stale';

const CONNECTION_ICONS: Record<ConnectionState, string> = {
  disconnected: '\u25CB', connecting: '\u25D4', connected: '\u25CF', error: '\u2716', stale: '\u25D2',
};

const CONNECTION_COLORS: Record<ConnectionState, string> = {
  disconnected: '#888888', connecting: '#eab308', connected: '#22c55e', error: '#ef4444', stale: '#eab308',
};

const MODE_LABELS: Record<BindingMode, string> = {
  coupled: 'Coupled', rest: 'REST', graphql: 'GraphQL', static: 'Static',
};

// --------------- Context ---------------

export interface BindingContextValue {
  binding: BindingConfig;
  connectionState: ConnectionState;
  readSignal(name: string): unknown;
  invoke(action: string, params?: Record<string, unknown>): Promise<void>;
}

let _currentBindingContext: BindingContextValue | null = null;

export function getBinding(): BindingContextValue {
  if (!_currentBindingContext) {
    throw new Error('getBinding must be called within a BindingProvider scope.');
  }
  return _currentBindingContext;
}

export function getBoundSignal(name: string): unknown {
  const ctx = getBinding();
  const signal = ctx.binding.signalMap[name];
  return signal?.get();
}

// --------------- Props ---------------

export interface BindingProviderProps {
  binding: BindingConfig;
  connectionState?: ConnectionState;
  errorMessage?: string;
  showStatusBar?: boolean;
  statusBarPosition?: 'top' | 'bottom';
  showSignals?: boolean;
  width?: number;
  lastSync?: string;
  onRetry?: () => void;
  onDisconnect?: () => void;
  children?: import('@nativescript/core').View[];
}

// --------------- Component ---------------

export function createBindingProvider(props: BindingProviderProps): StackLayout {
  const {
    binding,
    connectionState = 'disconnected',
    errorMessage,
    showStatusBar = true,
    statusBarPosition = 'top',
    showSignals = false,
    width,
    lastSync,
    onRetry,
    onDisconnect,
    children = [],
  } = props;

  // Set context
  _currentBindingContext = {
    binding,
    connectionState,
    readSignal: (name: string) => binding.signalMap[name]?.get(),
    invoke: async (action, params) => {
      // Placeholder for action invocation
      console.log(`BindingProvider invoke: ${action}`, params);
    },
  };

  const container = new StackLayout();
  container.className = 'clef-binding-provider';
  if (width) container.width = width;

  // Status bar
  const createStatusBar = (): StackLayout => {
    const bar = new StackLayout();
    bar.orientation = 'horizontal';
    bar.padding = 4;
    bar.backgroundColor = new Color('#1a1a2e');

    // Connection icon
    const icon = new Label();
    icon.text = CONNECTION_ICONS[connectionState];
    icon.color = new Color(CONNECTION_COLORS[connectionState]);
    icon.marginRight = 4;
    bar.addChild(icon);

    // Mode label
    const mode = new Label();
    mode.text = MODE_LABELS[binding.mode];
    mode.fontWeight = 'bold';
    mode.color = new Color('#ffffff');
    mode.marginRight = 8;
    bar.addChild(mode);

    // Concept name
    const concept = new Label();
    concept.text = String(binding.concept);
    concept.color = new Color('#06b6d4');
    bar.addChild(concept);

    // Status text
    const status = new Label();
    status.text = ` [${connectionState}]`;
    status.color = new Color(CONNECTION_COLORS[connectionState]);
    bar.addChild(status);

    // Error message
    if (connectionState === 'error' && errorMessage) {
      const errLabel = new Label();
      errLabel.text = ` \u2716 ${errorMessage}`;
      errLabel.color = new Color('#ef4444');
      bar.addChild(errLabel);
    }

    // Last sync
    if (lastSync) {
      const syncLabel = new Label();
      syncLabel.text = ` sync: ${lastSync}`;
      syncLabel.opacity = 0.5;
      syncLabel.fontSize = 11;
      bar.addChild(syncLabel);
    }

    return bar;
  };

  if (showStatusBar && statusBarPosition === 'top') {
    container.addChild(createStatusBar());
  }

  // Signal display
  if (showSignals && Object.keys(binding.signalMap).length > 0) {
    const signalContainer = new StackLayout();
    const signalHeader = new Label();
    signalHeader.text = 'Signals:';
    signalHeader.opacity = 0.5;
    signalContainer.addChild(signalHeader);

    Object.entries(binding.signalMap).forEach(([name, signal]) => {
      const row = new StackLayout();
      row.orientation = 'horizontal';
      const nameLabel = new Label();
      nameLabel.text = `  ${name}: `;
      nameLabel.opacity = 0.5;
      row.addChild(nameLabel);
      const valueLabel = new Label();
      valueLabel.text = JSON.stringify(signal.get());
      valueLabel.color = new Color('#06b6d4');
      row.addChild(valueLabel);
      signalContainer.addChild(row);
    });

    container.addChild(signalContainer);
  }

  // Children
  for (const child of children) {
    container.addChild(child);
  }

  if (showStatusBar && statusBarPosition === 'bottom') {
    container.addChild(createStatusBar());
  }

  return container;
}

export default createBindingProvider;

File: surface/widgets/compose/components/BindingProvider.kt
// ============================================================
// Clef Surface Compose Widget — BindingProvider
//
// Manages Clef Surface concept binding in Compose context.
// Provides signal-based data flow via CompositionLocal and
// renders connection status indicators using Material 3.
// ============================================================

package clef.surface.compose.components

import androidx.compose.foundation.layout.*
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp

// --------------- Types ---------------

enum class ConnectionState(val icon: String, val color: Color) {
    DISCONNECTED("\u25CB", Color.Gray),
    CONNECTING("\u25D4", Color(0xFFFFC107)),
    CONNECTED("\u25CF", Color(0xFF4CAF50)),
    ERROR("\u2716", Color(0xFFF44336)),
    STALE("\u25D2", Color(0xFFFFC107)),
}

enum class BindingMode(val label: String, val icon: String) {
    COUPLED("Coupled", "\u21C4"),
    REST("REST", "\u21BB"),
    GRAPHQL("GraphQL", "\u25C6"),
    STATIC("Static", "\u25A0"),
}

data class BindingConfig(
    val concept: String,
    val mode: BindingMode,
    val endpoint: String? = null,
    val signalMap: Map<String, Any> = emptyMap(),
)

// --------------- Context ---------------

data class BindingContextValue(
    val binding: BindingConfig,
    val connectionState: ConnectionState,
    val readSignal: (String) -> Any?,
    val invoke: suspend (String, Map<String, Any>?) -> Unit,
)

val LocalBinding = compositionLocalOf<BindingContextValue?> { null }

@Composable
fun rememberBinding(): BindingContextValue {
    return LocalBinding.current
        ?: error("rememberBinding must be used within a BindingProvider.")
}

// --------------- Component ---------------

@Composable
fun BindingProvider(
    binding: BindingConfig,
    connectionState: ConnectionState = ConnectionState.DISCONNECTED,
    errorMessage: String? = null,
    showStatusBar: Boolean = true,
    statusBarPosition: String = "bottom",
    showSignals: Boolean = false,
    lastSync: Long? = null,
    accentColor: Color = MaterialTheme.colorScheme.primary,
    onInvoke: (suspend (String, Map<String, Any>?) -> Unit)? = null,
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val contextValue = remember(binding, connectionState) {
        BindingContextValue(
            binding = binding,
            connectionState = connectionState,
            readSignal = { name -> binding.signalMap[name] },
            invoke = { action, params -> onInvoke?.invoke(action, params) },
        )
    }

    CompositionLocalProvider(LocalBinding provides contextValue) {
        Column(modifier = modifier) {
            if (showStatusBar && statusBarPosition == "top") {
                BindingStatusBar(binding, connectionState, errorMessage, lastSync, accentColor)
            }

            if (showSignals && binding.signalMap.isNotEmpty()) {
                SignalDebugPanel(binding)
            }

            content()

            if (showStatusBar && statusBarPosition == "bottom") {
                BindingStatusBar(binding, connectionState, errorMessage, lastSync, accentColor)
            }
        }
    }
}

@Composable
private fun BindingStatusBar(
    binding: BindingConfig,
    state: ConnectionState,
    errorMessage: String?,
    lastSync: Long?,
    accentColor: Color,
) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        modifier = Modifier.fillMaxWidth(),
    ) {
        Text(
            text = state.icon,
            color = state.color,
            style = MaterialTheme.typography.labelSmall,
        )
        Text(
            text = binding.concept,
            color = accentColor,
            style = MaterialTheme.typography.labelSmall,
        )
        Text(
            text = "${binding.mode.icon} ${binding.mode.label}",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        binding.endpoint?.let {
            Text(
                text = "\u2192 $it",
                style = MaterialTheme.typography.labelSmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
            )
        }
        if (state == ConnectionState.ERROR && errorMessage != null) {
            Text(
                text = errorMessage,
                color = Color(0xFFF44336),
                style = MaterialTheme.typography.labelSmall,
            )
        }
        Text(
            text = "[${binding.signalMap.size} signals]",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
    }
}

@Composable
private fun SignalDebugPanel(binding: BindingConfig) {
    Column {
        Text(
            text = "Signals:",
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        binding.signalMap.forEach { (name, value) ->
            Row(modifier = Modifier.padding(start = 8.dp)) {
                Text(
                    text = "\u25B8 $name: ",
                    color = MaterialTheme.colorScheme.tertiary,
                    style = MaterialTheme.typography.labelSmall,
                )
                Text(
                    text = value.toString(),
                    style = MaterialTheme.typography.labelSmall,
                )
            }
        }
    }
}
